tools/asinfo/asinfo.1.in
.\" Copyright (c) 2017 Edgar Kaziakhmedov <edgar.kaziakhmedov@virtuozzo.com>
.\" Copyright (c) 1996-2017 The strace developers.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. The name of the author may not be used to endorse or promote products
.\"    derived from this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
.\" THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

.\" Required option.
.de OR
.  ie \\n(.$-1 \
.    RI "\fB\\$1\fP" "\ \\$2"
.  el \
.    BR "\\$1"
..
.TH ASINFO 1 "@ASINFO_MANPAGE_DATE@" "strace package @VERSION@"
.SH NAME
asinfo \- advanced system call information tool
.SH SYNOPSIS
.SY asinfo
.BR "" \fR[{
.OR \-\-set\-arch arch
.BR "" |
.OR \-\-get\-arch
.BR "" }
.OR \fR[\fP\-\-set\-abi abi
.BR "" |
.OR \-\-list\-abi\fR]\fR]
.BR "" {
.BR "" {
.OR \-\-get\-sname expr
.BR "" |
.OR \-\-get\-snum expr
.BR "" }
.OP \-\-nargs
.BR "" }
.OP "\-\-raw"
.YS
.SY asinfo
.BR "" {
.OR \-\-set\-arch arch
.BR "" |
.OR \-\-get\-arch
.BR "" |
.OR \-\-list\-arch
.BR "" }
.BD "" [
.OR \fR[\fP\-\-set\-abi abi
.BR "" |
.OR \-\-list\-abi\fR]
.OP "\-\-raw"

.SH DESCRIPTION
.B asinfo
is a useful tool, which aimed to provide information about system calls,
architectures, and application binary interfaces (ABIs). In the simplest
case it provides mapping from system call name to number and reverse.
The main advantage of tool is it can work in single/multi arch modes with
the opportunity to show discrepancies in system call characteristics.
Also, the single arch mode allows program to take a guess about the
current architecture and ABI, if they are not specified. Furthermore,
.B asinfo
provides convenient filtering for selecting system calls.

.SH OPTIONS
.SS "Architecture parameters"
.TP 7
.BI "\-\-set\-arch " arch
Specify architecture/architectures manually. The format of the
.I arch
expression is:
.RS 9
.IP
\fIarch1\/\fR[\fB,\fIarch2\/\fR]...
.RE
.IP
.TP
.B \-\-get\-arch
Select achitecture based on the current machine.
.TP
.B \-\-list-arch
Print out all supported architectures.
Combined use with any ABI option is permitted.
.SS "ABI parameters"
.TP 7
.BI "\-\-set\-abi " abi
Specify ABI/ABIs manually. The format of the experession is:
.RS 9
.IP
\fIabi1\/\fR[\fB,\fIabi2\/\fR]...
.RE
.IP
.IP
Note that ABI should be selected for each corresponding architecture.
In addition, the special value
.B all
allows to choose all ABIs for the respective architecture.
.TP
.B "\-\-list\-abi "
Select all ABIs for the chosen architecture/architectures.
.IP
If ABI parameters are not used and only single architecture is selected, tool
will take a guess about ABI based on the strace package build.
.SS "System call parameters"
.TP 7
.BI "\-\-get\-sname " expr
Select system calls that satisfy a filtering expression
.I
expr
and print out name of system calls and numbers for each architecture/ABI.
.TP
.BI "\-\-get\-snum " expr
Select system calls that satisfy a filtering expression
.I
expr
and print out number of system calls and names for each architecture/ABI.
.TP
.B \-\-nargs
Switch the second output system call characteristic to number of arguments.
.SS Output formatting
.TP 7
.B "\-\-raw"
Reset alignment and remove titles, use ';' as a delimiter.
.SS Miscellaneous
.TP 7
.B \-h
Print the help summary.
.TP
.B \-v
Print the version number.

.SH "FILTERING EXPRESSION"
A filtering expression is a pattern that describes a set of syscall names,
syscall numbers, and syscall group. The format of the expression is:
.RS 2
.IP
[\fB!\fR][\fB?\fR]\,\fIvalue1\/\fR[\fB,\fR[\fB?\fR]\,\fIvalue2\/\fR]...
.RE
.LP
where
.I value
is a symbol or number. Using an exclamation mark negates the set of values.
For example,
.BR \fIvalue\fR = write
means print strictly the write system call.  By contrast,
.BR \fIvalue\fR = !write
means to dump every system call except write. Question mark before the
syscall qualification allows suppression of error in case no syscalls matched
the qualification provided, that can be particularly useful in multiarch mode,
when system call is not presented in all selected architectures. In addition,
the special values
.B all
and
.B none
have the obvious meanings.
.LP
Note that some shells use the exclamation point for history
expansion even inside quoted arguments.  If so, you must escape
the exclamation point with a backslash.
.SS "Strict match"
.TP 7
.B \fIvalue\fR=\,\fIset\fR
Print out only the specified set of system calls. For example,
.BR \fIvalue\fR = open,close,read,write
means to only show those four system calls.
.SS "Regex match"
.TP 7
.B \fIvalue\fR=/\,\fIregex\fR
Show only those system calls that match the
.IR regex .
You can use
.B POSIX
Extended Regular Expression syntax (see
.BR regex (7)).
.SS "Class match"
.TP 7
.BR \fIvalue\fR = %file
.TQ
.BR \fIvalue\fR  = file " (deprecated)"
Show all system calls which take a file name as an argument.  You
can think of this as an abbreviation for
.BR \fIvalue\fR = open,stat,chmod,unlink,...
Furthermore, using the abbreviation will ensure that you don't
accidentally forget to include a call like
.B lstat
in the list.
.PP
.BR \fIvalue\fR = %process
.TQ
.BR \fIvalue\fR = process " (deprecated)"
Show all system calls which involve process management.
.PP
.BR \fIvalue\fR = %network
.TQ
.BR \fIvalue\fR = network " (deprecated)"
Show all the network related system calls.
.PP
.BR \fIvalue\fR = %signal
.TQ
.BR \fIvalue\fR = signal " (deprecated)"
Show all signal related system calls.
.PP
.BR \fIvalue\fR = %ipc
.TQ
.BR \fIvalue\fR = ipc " (deprecated)"
Show all IPC related system calls.
.PP
.BR \fIvalue\fR = %desc
.TQ
.BR \fIvalue\fR = desc " (deprecated)"
Show all file descriptor related system calls.
.PP
.BR \fIvalue\fR = %memory
.TQ
.BR \fIvalue\fR = memory " (deprecated)"
Show all memory mapping related system calls.
.TP
.BR \fIvalue\fR = %stat
Show stat syscall variants.
.TP
.BR \fIvalue\fR = %lstat
Show lstat syscall variants.
.TP
.BR \fIvalue\fR = %fstat
Show fstat and fstatat syscall variants.
.TP
.BR \fIvalue\fR = %%stat
Show syscalls used for requesting file status (stat, lstat, fstat, fstatat,
statx, and their variants).
.TP
.BR \fIvalue\fR = %statfs
Show statfs, statfs64, statvfs, osf_statfs, and osf_statfs64 system calls.
The same effect can be achieved with
.BR \fIvalue\fR = /^(.*_)?statv?fs
regular expression.
.TP
.BR \fIvalue\fR = %fstatfs
Show fstatfs, fstatfs64, fstatvfs, osf_fstatfs, and osf_fstatfs64 system calls.
The same effect can be achieved with
.BR \fIvalue\fR = /fstatv?fs
regular expression.
.TP
.BR \fIvalue\fR = %%statfs
Show syscalls related to file system statistics (statfs-like, fstatfs-like,
and ustat).  The same effect can be achieved with
.BR \fIvalue\fR = /statv?fs|fsstat|ustat
regular expression.

.SH "EXIT STATUS"
On success,
.B asinfo
returns 0. Otherwise, in case of wrong input or no matches found, 1.

.SH "REPORTING BUGS"
Problems with
.B asinfo
should be reported to the
.B strace
mailing list at <strace-devel@lists.sourceforge.net>.

.SH "SEE ALSO"
.BR strace (1)


NEWS
Noteworthy changes in release ?.?? (????-??-??)
===============================================

* New tools
  * Added asinfo tool that is purposed to provide information about
    system calls and architectures.

* Improvements
  * Implemented decoding of KVM_CHECK_EXTENSION ioctl command.
  * Implemented decoding of SOL_XDP socket option names.
  * Implemented decoding of AF_PACKET protocols in socket syscall.
  * Implemented decoding of AF_AX25 and AF_X25 addresses, protocols,
    and options.
  * Implemented stack trace printing on receiving signals.
  * Enhanced xlat styles support configured by -X option.
  * Enhanced decoding of IFLA_*, NETCONFA_*, and SMC_DIAG_* netlink attributes.
  * Wired up io_pgetevents syscall on arm, sparc, and sparc64.
  * Updated lists of AF_*, AUDIT_*, BPF_*, IFLA_*, IN_*, IPSTATS_*,
    IPV4_DEVCONF_*, NT_*, SCTP_*, SO_*, SOCK_*, SOL_*, and V4L2_* constants.
  * Updated lists of ioctl commands from Linux 4.19.

* Bug fixes
  * Fixed decoding of bit sets in evdev ioctl commands.
  * Fixed decoding of EVIOCGBIT(0, ...) ioctl commands.
  * Fixed test suite for the case when mlockall and unimplemented syscalls
    are explicitly disabled by a seccomp filter.

Noteworthy changes in release 4.24 (2018-08-14)
===============================================

* Improvements
  * Implemented decoding of kvm vcpu exit reason (-e kvm=vcpu option),
    requires Linux kernel version 4.16.0 or higher.
  * Implemented decoding of KVM_SET_CPUID2 and KVM_GET_*_CPUID ioctl commands.
  * Enhanced decoding of corner cases of SO_LINGER, SO_PEERCRED,
    and PACKET_STATISTICS getsockopt options.
  * Wired up io_pgetevents and rseq syscalls on hppa, microblaze, mips, powerpc,
    and s390.
  * Updated lists of AUDIT_*, BPF_*, IFA_*, IFLA_*, RTA_*, RTPROT_*, TCP_*,
    TRAP_*, and UDP_* constants.
  * Updated lists of ioctl commands from Linux 4.18.

* Bug fixes
  * Fixed decoding of Q_XGETQSTAT on non-native personalities.
  * Fixed race condition in aio test.
  * Fixed build with Bionic libc.
  * Fixed build with glibc-2.28.
  * Fixed build in exotic systemd-nspawn based environments.
  * Fixed build on s390 with kernel headers that do not contain Linux commit
    v3.10-rc1~201^2~11.

Noteworthy changes in release 4.23 (2018-06-14)
===============================================

* Changes in behaviour
  * On x32 personality, 64-bit syscalls (such as readv) are now shown with "#64"
    suffix instead of "64:" prefix.  Unlike the old names, these new names
    can be used in syscall qualification expressions.
  * Changed error number output format: error numbers are consistently
    printed as "-1 ECONST (Error description)"
    for known errors and as "-1 (errno 123)" for unknown errors,
    regardless of -e raw qualification settings.  One can specify
    -X raw for displaying raw error numbers unconditionally.
  * Unfetchable addresses inside arrays are now printed as comments.
  * Obsolete IA-32 mode syscall names are no longer printed on ia64.

* Improvements
  * Implemented libdw backend for -k option, configured at build time
    using --with-libdw option.
    Whether -k option is compiled is now configured at build time
    using --enable-stacktrace option.
  * Added -X option for configuring xlat output formatting (addresses
    Debian bug #692915).
  * Added support for personality designation ("64", "32", or "x32") to syscall
    qualifications in -e trace expressions.
  * Implemented injection of syscalls with no side effects as an alternative
    to injection of an invalid syscall (-e inject=SET:syscall= expression).
  * Improved support for reproducible builds (addresses Debian bug #896016).
  * Implemented decoding of BPF_PROG_QUERY and BPF_RAW_TRACEPOINT_OPEN bpf
    syscall commands.
  * Implemented decoding of INOTIFY_IOC_SETNEXTWD and PERF_EVENT_IOC_* ioctl
    commands.
  * Implemented decoding of PR_GET_SPECULATION_CTRL and PR_SET_SPECULATION_CTRL
    prctl syscall options.
  * Enhanced decoding of bpf, getsockopt, setsockopt, and socket syscalls.
  * Enhanced decoding of NETLINK_KOBJECT_UEVENT and NETLINK_ROUTE protocols.
  * Enhanced decoding of *_DIAG_MEMINFO netlink attributes.
  * Enhanced decoding of BTRFS_*, FS_IOC_*, SIOCGIFHWADDR, and SIOCSIFHWADDR
    ioctl commands.
  * Enhanced decoding of AF_BLUETOOTH socket addresses.
  * Implemented decoding of io_pgetevent syscall.
  * Wired up rseq syscall.
  * Updated lists of ADJ_*, BPF_*, BPF_F_*, BPF_PROG_TYPE_*, ETH_P_*, FPE_*,
    IFF_*, IPPROTO_*, MAP_*, MS_*, MSG_*, NETLINK_*, PACKET_*, PROT_*, SCTP_*,
    SECCOMP_FILTER_FLAG_*, SEGV_*, SEM_*, SHM_*, SW_*, V4L2_CID_*,
    V4L2_CTRL_CLASS_*, and V4L2_PIX_FMT_* constants.
  * Updated lists of ioctl commands from Linux 4.17.

* Bug fixes
  * Fixed decoding of msgtyp argument of msgrcv syscall.
  * Fixed fallback definitions for F_*, MCL_*, MSG_ZEROCOPY, SI_*,
    and SOL_SOCKET constants.
  * Fixed -e trace=%pure expression on mips.

Noteworthy changes in release 4.22 (2018-04-05)
===============================================

* Changes in infrastructure
  * The mailing list was moved to strace-devel@lists.strace.io.
  * SourceForge git repository
    (https://sourceforge.net/p/strace/code/ci/master/tree/) is no longer
    maintained.  Please use GitLab (https://gitlab.com/strace/strace/)
    or GitHub (https://github.com/strace/strace/) repository instead.

* Changes in behaviour
  * When printing string data, '8' and '9' following an octal escape sequence
    no longer lead to unabbreviated octal escape sequence output, as they used
    to do.
  * Angle brackets are now printed as octal number escape sequences
    in the output of paths associated with file descriptors.
  * Data dump for write syscalls (-e write) is performed regardless
    of successfulness of these syscall (addresses Debian bug #436284).
  * -r and -t options are now affect output independently.  Previously,

tools/asinfo/asinfo.c
/*
 * The asinfo main source. The asinfo tool is purposed to operate
 * with system calls and provide information about it.
 *
 * Copyright (c) 2017 Edgar A. Kaziakhmedov <edgar.kaziakhmedov@virtuozzo.com>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the author may not be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif
#include <errno.h>
#include <ctype.h>
#include <limits.h>
#include <stdarg.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "arch_interface.h"
#include "dispatchers.h"
#include "error_interface.h"
#include "error_prints.h"
#include "macros.h"
#include "request_msgs.h"
#include "syscall_interface.h"
#include "xmalloc.h"

#ifndef HAVE_PROGRAM_INVOCATION_NAME
char *program_invocation_name;
#endif

static void
usage(void)
{
	puts(
		"usage: asinfo (--set-arch arch | --get-arch | --list-arch)\n"
		"              [--set-abi abi | --list-abi] [--raw]\n"
		"   or: asinfo [(--set-arch arch | --get-arch) [--set-abi abi | --list-abi]]\n"
		"              ((--get-sname expr | --get-snum expr) [--nargs]) [--raw]\n"
		"\n"
		"Architecture:\n"
		"  --set-arch arch  use architecture ARCH for further work\n"
		"                   argument format: arch1,arch2,...\n"
		"  --get-arch       use architecture returned by uname for further work\n"
		"  --list-arch      print out all architectures supported by strace\n"
		"                   (combined use list-arch and any ABI option is permitted)\n"
		"\n"
		"ABI:\n"
		"  --set-abi abi    use application binary interface ABI for further work\n"
		"                   ('all' can be used as ABI to use all compatible personalities\n"
		"                   for corresponding architecture)\n"
		"                   argument format: abi1,abi2,...\n"
		"  --list-abi       use all ABIs for specified architecture\n"
		"\n"
		"System call:\n"
		"  --get-sname expr print all system calls that satisfy a filtering expression:\n"
		"                   [!]all or [!][?]val1[,[?]val2]...\n"
		"                   with the following format:\n"
		"                   | N | syscall name | snum1 | snum2 | ...\n"
		"  --get-snum expr  print all system calls that satisfy a filtering expression:\n"
		"                   [!]all or [!][?]val1[,[?]val2]...\n"
		"                   with the following format:\n"
		"                   | N | syscall number | sname1 | sname2 | ...\n"
		"  --nargs          change output format as follows:\n"
		"                   | N | syscall name or number | nargs1 | sname2 | ...\n"
		"\n"
		"Output formatting:\n"
		"  --raw            reset alignment and remove titles, use ';' as a delimiter\n"
		"\n"
		"Miscellaneous:\n"
		"  -h               print help message\n"
		"  -v               print version");
	exit(0);
}

static void
print_version(void)
{
	printf("asinfo (%s package) -- version %s\n"
	       "Copyright (c) 1991-%s The strace developers <%s>.\n"
	       "This is free software; see the source for copying conditions.  There is NO\n"
	       "warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n",
	       PACKAGE_NAME, PACKAGE_VERSION, COPYRIGHT_YEAR, PACKAGE_URL);
	exit(0);
}

void
die(void)
{
	exit(1);
}

static int
is_more1bit(unsigned int num)
{
	return !(num & (num - 1));
}

static unsigned
strpar2req(char *option)
{
	/* Convertion table to store string with options */
	static const char *options[] = {
		[SD_REQ_GET_SNAME_BIT]	= "--get-sname",
		[SD_REQ_GET_SNUM_BIT]	= "--get-snum",
		[SD_REQ_NARGS_BIT]	= "--nargs",
		[AD_REQ_SET_ARCH_BIT]	= "--set-arch",
		[AD_REQ_GET_ARCH_BIT]	= "--get-arch",
		[AD_REQ_LIST_ARCH_BIT]	= "--list-arch",
		[ABD_REQ_SET_ABI_BIT]	= "--set-abi",
		[ABD_REQ_LIST_ABI_BIT]	= "--list-abi",
		[SERV_REQ_RAW_BIT]	= "--raw",
		[SERV_REQ_HELP_BIT]	= "-h",
		[SERV_REQ_VERSION_BIT]	= "-v"
	};
	unsigned i;

	for (i = 0; i < ARRAY_SIZE(options); i++) {
		if (options[i] && strcmp(option, options[i]) == 0)
			return i;
	}
	return SERV_REQ_ERROR_BIT;
}

static char **
arg2list(char *argument)
{
	int i;
	int len = strlen(argument);
	int occur = 1;
	char **arg_list;

	for (i = 0; i < len; i++) {
		if (argument[i] == ',') {
			if (i == 0 || i == len - 1 || argument[i + 1] == ',')
				return NULL;
			occur++;
		}
	}
	arg_list = xcalloc(sizeof(*arg_list), occur + 1);
	for (i = 0; i < occur; i++) {
		arg_list[i] = argument;
		argument = strchr(argument, ',');
		if (argument) {
			*argument = '\0';
			argument++;
		}
	}
	return arg_list;
}

/* The purpose of this function is to convert input parameters to number with
   set bits, where each bit means specific work mode. Moreover, it checks input
   for correctness and outputs error messages in case of wrong input */
static unsigned
command_dispatcher(int argc, char *argv[], char **args[])
{
	int i;
	unsigned final_req = 0;
	unsigned temp_req = 0;
	int mult_arch = 0;
	int mult_abi = 0;
	unsigned non_req_arg = AD_REQ_GET_ARCH	| AD_REQ_LIST_ARCH	|
			       ABD_REQ_LIST_ABI	| SD_REQ_NARGS		|
			       SERV_REQ_RAW;

	if (!program_invocation_name || !*program_invocation_name) {
		static char name[] = "asinfo";
		program_invocation_name =
			(argv[0] && *argv[0]) ? argv[0] : name;
	}

	/* Try to find help or version parameter first */
	for (i = 1; i < argc; i++) {
		if (strpar2req(argv[i]) == SERV_REQ_HELP_BIT)
			usage();
		if (strpar2req(argv[i]) == SERV_REQ_VERSION_BIT)
			print_version();
	}
	/* For now, is is necessary to convert string parameter to number of
	   request and make basic check */
	for (i = 1; i < argc; i++) {
		if ((temp_req = strpar2req(argv[i])) == SERV_REQ_ERROR_BIT)
			error_msg_and_help("unrecognized option '%s'",
					   argv[i]);
		if (final_req & 1 << temp_req)
			error_msg_and_help("parameter '%s' has been used "
					   "more than once", argv[i]);
		if (!((1 << temp_req) & non_req_arg) &&
		     (i + 1 >= argc || strlen(argv[i + 1]) == 0 ||
		      strpar2req(argv[i + 1]) != SERV_REQ_ERROR_BIT))
			error_msg_and_help("parameter '%s' requires "
					   "argument", argv[i]);
		final_req |= 1 << temp_req;
		if (!((1 << temp_req) & non_req_arg)) {
			if ((1 << temp_req) & SD_REQ_MASK) {
				args[temp_req] = &argv[i + 1];
				i++;
				continue;
			}
			if ((args[temp_req] = arg2list(argv[i + 1])) != NULL) {
				i++;
				continue;
			}
			error_msg_and_help("argument '%s' of '%s' parameter "
					   "has a wrong format",
					   argv[i + 1], argv[i]);
		}
	}
	/* Count our multuarchness */
	if (args[AD_REQ_SET_ARCH_BIT])
		while (args[AD_REQ_SET_ARCH_BIT][mult_arch] != NULL)
			mult_arch++;
	if (args[ABD_REQ_SET_ABI_BIT])
		while (args[ABD_REQ_SET_ABI_BIT][mult_abi] != NULL)
			mult_abi++;
	/* final_req should be logically checked */
	/* More than one option from one request group couldn't be set */
	if ((is_more1bit(final_req & SD_REQ_MASK & ~SD_REQ_NARGS) == 0) ||
	    (is_more1bit(final_req & AD_REQ_MASK) == 0) ||
	    (is_more1bit(final_req & ABD_REQ_MASK) == 0))
		error_msg_and_help("exclusive parameters");
	/* Check on mutually exclusive options chain */
	/* If at least one syscall option has been typed, therefore
	   arch_options couldn't be list-arch and
	   abi_option couldn't be list-abi */
	if ((final_req & SD_REQ_MASK) &&
	    (((final_req & AD_REQ_MASK) && (final_req & AD_REQ_LIST_ARCH))))
		error_msg_and_help("wrong parameters");
	/* list-arch couldn't be used with any abi options */
	if ((final_req & AD_REQ_LIST_ARCH) &&
	    (final_req & ABD_REQ_MASK))
		error_msg_and_help("'--list-arch' cannot be used with any "
				   "ABI parameters");
	/* ABI requests could be used just in a combination with arch
	   requests */
	if ((final_req & ABD_REQ_MASK) &&
	    !(final_req & AD_REQ_MASK))
		error_msg_and_help("ABI parameters could be used only with "
				   "architecture parameter");
	/* set-abi must be used in case of multiple arch */
	if ((mult_arch > 1) && !(final_req & ABD_REQ_MASK))
		error_msg_and_help("ABI modes cannot be automatically "
				   "detected for multiple "
				   "architectures");
	/* set-abi and set-arch have to take the same number of args */
	if ((final_req & AD_REQ_SET_ARCH) && (final_req & ABD_REQ_SET_ABI) &&
	    (mult_arch != mult_abi))
		error_msg_and_help("each architecture needs respective "
				   "ABI mode, and vice versa");
	/* --nargs cannot be used alone */
	if ((final_req & SD_REQ_NARGS) &&
	    !(final_req & SD_REQ_MASK & ~SD_REQ_NARGS))
		error_msg_and_help("first set main output syscall "
				   "characteristics");
	/* raw should not be single */
	if (final_req == SERV_REQ_RAW)
		error_msg_and_help("raw data implies existing data");
	return final_req;
}

static char *
seek_sc_arg(char **input_args[])
{
	int i;

	for (i = SD_REQ_GET_SNAME_BIT; i < SYSCALL_REQ_BIT_LAST; i++)
		if (input_args[i] != NULL)
			return input_args[i][0];
	return NULL;
}

int
main(int argc, char *argv[])
{
	ARCH_LIST_DEFINE(arch_list);
	SYSCALL_LIST_DEFINE(sc_list);
	/* This array is purposed to store arguments for options in the
	   most convenient way */
	char ***in_args = xcalloc(sizeof(*in_args), REQ_LAST_BIT);
	unsigned reqs;

	/* command_dispatcher turn */
	reqs = command_dispatcher(argc, argv, in_args);
	if (reqs == 0)
		error_msg_and_help("must have OPTIONS");

	/* arch_dispatcher turn */
	arch_list = arch_dispatcher(reqs, in_args[AD_REQ_SET_ARCH_BIT]);
	if (es_error(al_err(arch_list)))
		perror_msg_and_die("%s", es_get_serror(al_err(arch_list)));
	/* abi_dispatcher turn */
	abi_dispatcher(arch_list, reqs, in_args[ABD_REQ_SET_ABI_BIT]);
	if (es_error(al_err(arch_list)))
		perror_msg_and_die("%s", es_get_serror(al_err(arch_list)));
	/* syscall_dispatcher turn */
	sc_list = syscall_dispatcher(arch_list, reqs, seek_sc_arg(in_args));
	if (es_error(ss_err(sc_list)))
		perror_msg_and_die("%s", es_get_serror(ss_err(sc_list)));
	/* If we want to get info about only architectures thus we print out
	   architectures, otherwise system calls */
	if (!(reqs & SD_REQ_MASK))
		al_dump(arch_list, reqs & SERV_REQ_RAW);
	else
		ss_dump(sc_list, reqs & SERV_REQ_RAW);
	return 0;
}
